File: src/main/java/io/github/boogiemonster1o1/palpebratingpesl/objects/UUIDObject.java
package io.github.boogiemonster1o1.palpebratingpesl.objects;

import java.util.UUID;

import javax.annotation.Nonnull;

import io.github.boogiemonster1o1.palpebratingpesl.util.FunctionObjects;
import p0nki.pesl.api.object.BuiltinMapLikeObject;
import p0nki.pesl.api.object.NumberObject;
import p0nki.pesl.api.object.StringObject;

public class UUIDObject extends BuiltinMapLikeObject {
	private final UUID uuid;

	public UUIDObject(UUID uuid) {
		super("uuid");
		this.uuid = uuid;
		this.put("getMostSigBits", FunctionObjects.of(this, (args) -> new NumberObject(this.uuid.getMostSignificantBits())));
		this.put("getLeastSigBits", FunctionObjects.of(this, (args) -> new NumberObject(this.uuid.getLeastSignificantBits())));
		this.put("toString", FunctionObjects.of(this, (args) -> new StringObject(this.uuid.toString())));
	}

	@Nonnull
	@Override
	public String castToString() {
		return this.uuid.toString();
	}
}

File: src/main/kotlin/io/github/boogiemonster1o1/palpebratingpesl/objects/UuidObject.kt
package io.github.boogiemonster1o1.palpebratingpesl.objects

import p0nki.pesl.api.`object`.*
import java.util.*

class UuidObject(val uuid: UUID) : BuiltinMapLikeObject("uuid") {
	init {
		this.put("asString", FunctionObject.of(false) {
			StringObject(this.uuid.toString())
		})
		this.put("getMostSigBits", FunctionObject.of(false) {
			NumberObject(this.uuid.mostSignificantBits.toDouble())
		})
		this.put("getLeastSigBits", FunctionObject.of(false) {
			NumberObject(this.uuid.leastSignificantBits.toDouble())
		})
	}

	override fun equals(other: Any?): Boolean {
		if (other is UuidObject) {
			return other.uuid == this.uuid
		}
		return false
	}

	override fun compareEquals(`object`: PESLObject): Boolean {
		return this == `object`
	}

	override fun hashCode(): Int {
		return uuid.hashCode()
	}
}
